network/message/authorization.go

```diff
@@ -0,0 +1,375 @@
+)
+
+// MsgAuthConfig contains authorization information for a specific flow message. The authorization
+// is represented as a map from network channel -> list of all roles allowed to send the message on
+// the channel.
+type MsgAuthConfig struct {
+	// Name is the string representation of the message type.
+	Name string
+	// Type is a func that returns a new instance of message type.
+	Type func() interface{}
+	// Config is the mapping of network channel to list of authorized flow roles.
+	Config map[channels.Channel]flow.RoleList
+}
+
+// IsAuthorized checks if the specified role is authorized to send the message on the provided channel and
+// asserts that the message is authorized to be sent on the channel.
+// Expected error returns during normal operations:
+//  * ErrUnauthorizedMessageOnChannel: the channel is not included in the message's list of authorized channels
+//  * ErrUnauthorizedRole: the role is not included in the message's list of authorized roles for the provided channel
+func (m MsgAuthConfig) IsAuthorized(role flow.Role, channel channels.Channel) error {
+	authorizedRoles, ok := m.Config[channel]
+	if !ok {
+		return ErrUnauthorizedMessageOnChannel
+	}
+
+	if !authorizedRoles.Contains(role) {
+		return ErrUnauthorizedRole
+	}
+
+	return nil
+}
+
+var (
+	ErrUnknownMsgType               = errors.New("could not get authorization config for unknown message type")
+	ErrUnauthorizedMessageOnChannel = errors.New("message is not authorized to be sent on channel")
+	ErrUnauthorizedRole             = errors.New("sender role not authorized to send message on channel")
+	AuthorizationConfigs            map[string]MsgAuthConfig
+)
+
+func initializeMessageAuthConfigsMap() {
+	AuthorizationConfigs = make(map[string]MsgAuthConfig)
+
+	// consensus
+	AuthorizationConfigs[BlockProposal] = MsgAuthConfig{
+		Name: BlockProposal,
+		Type: func() interface{} {
+			return new(messages.BlockProposal)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.ConsensusCommittee: {flow.RoleConsensus},
+			channels.PushBlocks:         {flow.RoleConsensus}, // channel alias ReceiveBlocks = PushBlocks
+		},
+	}
+	AuthorizationConfigs[BlockVote] = MsgAuthConfig{
+		Name: BlockVote,
+		Type: func() interface{} {
+			return new(messages.BlockVote)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.ConsensusCommittee: {flow.RoleConsensus},
+		},
+	}
+
+	// protocol state sync
+	AuthorizationConfigs[SyncRequest] = MsgAuthConfig{
+		Name: SyncRequest,
+		Type: func() interface{} {
+			return new(messages.SyncRequest)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.SyncCommittee:     flow.Roles(),
+			channels.SyncClusterPrefix: {flow.RoleCollection},
+		},
+	}
+	AuthorizationConfigs[SyncResponse] = MsgAuthConfig{
+		Name: SyncResponse,
+		Type: func() interface{} {
+			return new(messages.SyncResponse)
+		},
+		Config: map[channels.Channel]flow.RoleList{
+			channels.SyncCommittee:     flow.Roles(),
```

Comment:
I believe this also should be just consensus nodes:
https://github.com/onflow/flow-go/blob/master/cmd/scaffold.go#L696-L696
```suggestion
			channels.SyncCommittee:     {flow.RoleConsensus},
```